sys/mips/mips/freebsd64_machdep.c

```diff
@@ -198,7 +200,224 @@ static void
+		kdb_enter(KDB_WHY_CHERI,
+		    "SIGPROT delivered outside sandbox");
+#endif
+
+	/*
+	 * If a thread is running sandboxed, we can't rely on $sp which may
+	 * not point at a valid stack in the ambient context, or even be
+	 * maliciously manipulated.  We must therefore always use the
+	 * alternative stack.  We are also therefore unable to tell whether we
+	 * are on the alternative stack, so must clear 'oonstack' here.
+	 *
+	 * XXXRW: This requires significant further thinking; however, the net
+	 * upshot is that it is not a good idea to do an object-capability
+	 * invoke() from a signal handler, as with so many other things in
+	 * life.
+	 */
+	if (cheri_is_sandboxed != 0)
+		oonstack = 0;
+#endif
+
+	/* save user context */
+	bzero(&sf, sizeof(sf));
+	sf.sf_uc.uc_sigmask = *mask;
+	sf.sf_uc.uc_stack.ss_sp = (__cheri_fromcap void *)td->td_sigstk.ss_sp;
+	sf.sf_uc.uc_stack.ss_size = td->td_sigstk.ss_size;
+	sf.sf_uc.uc_stack.ss_flags = td->td_sigstk.ss_flags;
+	sf.sf_uc.uc_mcontext.mc_onstack = (oonstack) ? 1 : 0;
+	sf.sf_uc.uc_mcontext.mc_pc = TRAPF_PC(regs);
+	sf.sf_uc.uc_mcontext.mullo = regs->mullo;
+	sf.sf_uc.uc_mcontext.mulhi = regs->mulhi;
+	sf.sf_uc.uc_mcontext.mc_tls =
+	    (__cheri_fromcap void *)td->td_md.md_tls;
+	sf.sf_uc.uc_mcontext.mc_regs[0] = UCONTEXT_MAGIC;  /* magic number */
+	bcopy((void *)&regs->ast, (void *)&sf.sf_uc.uc_mcontext.mc_regs[1],
+	    sizeof(sf.sf_uc.uc_mcontext.mc_regs) - sizeof(register_t));
+	sf.sf_uc.uc_mcontext.mc_fpused = td->td_md.md_flags & MDTD_FPUSED;
+#if defined(CPU_HAVEFPU)
+	if (sf.sf_uc.uc_mcontext.mc_fpused) {
+		/* if FPU has current state, save it first */
+		if (td == PCPU_GET(fpcurthread))
+			MipsSaveCurFPState(td);
+		bcopy((void *)&td->td_frame->f0,
+		    (void *)sf.sf_uc.uc_mcontext.mc_fpregs,
+		    sizeof(sf.sf_uc.uc_mcontext.mc_fpregs));
+	}
+#endif
+	/* XXXRW: sf.sf_uc.uc_mcontext.sr seems never to be set? */
+	sf.sf_uc.uc_mcontext.cause = regs->cause;
+
+	/* Allocate and validate space for the signal handler context. */
+	if ((td->td_pflags & TDP_ALTSTACK) != 0 && !oonstack &&
+	    SIGISMEMBER(psp->ps_sigonstack, sig)) {
+		sp = (vm_offset_t)((__cheri_addr vaddr_t)td->td_sigstk.ss_sp +
+		    td->td_sigstk.ss_size);
+	} else {
+#ifdef CPU_CHERI
+		/*
+		 * Signals delivered when a CHERI sandbox is present must be
+		 * delivered on the alternative stack rather than a local one.
+		 * If an alternative stack isn't present, then terminate or
+		 * risk leaking capabilities (and control) to the sandbox (or
+		 * just crashing the sandbox).
+		 */
+		if (cheri_is_sandboxed) {
+			mtx_unlock(&psp->ps_mtx);
+			printf("pid %d, tid %d: signal in sandbox without "
+			    "alternative stack defined\n", td->td_proc->p_pid,
+			    td->td_tid);
+			sigexit(td, SIGILL);
+			/* NOTREACHED */
+		}
+#endif
+		sp = (vm_offset_t)regs->sp;
+	}
+#ifdef CPU_CHERI
+	cp2_len = sizeof(*cfp);
+	sp -= cp2_len;
+	sp &= ~(CHERICAP_SIZE - 1);
```

Comment:
Probably copied code but I would use:
```suggestion
	sp = rounddown2(sp, CHERICAP_SIZE);
```

Reply:
Will do.